### PR TITLE
Cubical complex

### PR DESCRIPTION
**Pull request by JacquesOlivierLachaud**
This PR requires PR1079 of DGtal: https://github.com/DGtal-team/DGtal/pull/1079
It adds two tools to visualize 3d implicit surfaces. The first tool thickens the zero-level set to extract f=0, then collapses it and projects it on f=0. The second tool extends the implicit function f as a 4D function F, then extracts the zero levelset F=0, then collapses it in 4D, and projects it back in 3D on f=0.

### CODE REVIEW

**Comment by JacquesOlivierLachaud** [comment]
implicit-1

**Comment by kerautret** [comment]
@JacquesOlivierLachaud great news tools!
Sorry for the delay but I was waiting the main associated PR of DGtal. It looks that travis fails due to new format, could you merge from current master ? Thanks by advance!

**Comment by JacquesOlivierLachaud** [comment]
Le 12/01/2016 10:49, Bertrand Kerautret a écrit :
> @JacquesOlivierLachaud <https://github.com/JacquesOlivierLachaud> great
> news tools!
> Sorry for the delay but I was waiting the main associated PR of DGtal.
> It looks that travis fails due to new format, could you merge from
> current master ? Thanks by advance!
> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/DGtal-team/DGtalTools/pull/226#issuecomment-170854957>.
> 
Yes, I look at it this afternoon.
Jaco

File: visualisation/3dImplicitSurfaceExtractorBy4DExtension.cpp [hunk 1]
```diff
@@ -0,0 +1,671 @@
...
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ **/
+
+/**
+ * @file cubical-complex-collapse.cpp
+ * @ingroup Examples
```
Comment (kerautret) [inline visualisation/3dImplicitSurfaceExtractorBy4DExtension.cpp]: -> @ingroup Visualisation

File: visualisation/3dImplicitSurfaceExtractorBy4DExtension.cpp [hunk 2]
```diff
@@ -0,0 +1,671 @@
...
+ * @author Jacques-Olivier Lachaud (\c jacques-olivier.lachaud@univ-savoie.fr )
+ * Laboratory of Mathematics (CNRS, UMR 5127), University of Savoie, France
+ *
+ * @date 2015/08/28
+ *
+ * An example file named cubical-complex-collapse.cpp.
```
Comment (kerautret) [inline visualisation/3dImplicitSurfaceExtractorBy4DExtension.cpp]: remove 
(Ideally we could add a script for the DGtalTools template ;)

File: visualisation/3dImplicitSurfaceExtractorBy4DExtension.cpp [hunk 3]
```diff
@@ -0,0 +1,671 @@
...
+  typedef typename RealPoint3::Coordinate  Ring;
+  typedef typename Space4::RealPoint       RealPoint4;
+  typedef typename Space4::RealVector      RealVector4;
+  typedef RealPoint4                       RealPoint;
+  typedef RealVector4                      RealVector;
+
+  ImplicitSurface4DProductExtension( Clone<Polynomial3> poly )
+    : f( poly )
+  {
+    fx  = derivative<0>( f );
```
Comment (kerautret) [inline visualisation/3dImplicitSurfaceExtractorBy4DExtension.cpp]: it comes from the DEC class but I d'ont see the include .. (I see perhaps DrawWithDisplay3DModifier.h)

File: visualisation/3dImplicitSurfaceExtractorBy4DExtension.cpp [hunk 4]
```diff
@@ -0,0 +1,671 @@
...
+      RealPoint4 p  = dshape->embed( dp );
+      RealPoint3 p3 = RealPoint3( p[ 0 ], p[ 1 ], p[ 2 ] );
+      points.push_back( p3 );
+    }
+}
+    
+
+int main( int argc, char** argv )
+{
+  typedef int                               Integer;
```
Comment (kerautret) [inline visualisation/3dImplicitSurfaceExtractorBy4DExtension.cpp]: perhaps we can later define some namespace in DGtal to reduce the typef list ;)

File: visualisation/3dImplicitSurfaceExtractorBy4DExtension.cpp [hunk 5]
```diff
@@ -0,0 +1,671 @@
...
+  typedef CC4::CellMapConstIterator         CellMapConstIterator;
+  typedef CC4::Cells                        Cells4;
+
+  //-------------- parse command line ----------------------------------------------
+  namespace po = boost::program_options;
+  po::options_description general_opt("Allowed options are: ");
+  general_opt.add_options()
+    ("help,h", "display this message")
+    ("polynomial,p", po::value<string>(), "the implicit polynomial whose zero-level defines the shape of interest." )
+    ("minAABB,a",  po::value<double>()->default_value( -10.0 ), "the min value of the AABB bounding box (domain)" )
```
Comment (kerautret) [inline visualisation/3dImplicitSurfaceExtractorBy4DExtension.cpp]: -> the min value of the Axis Aligned Bounding Box (domain) ?

File: visualisation/3dImplicitSurfaceExtractorBy4DExtension.cpp [hunk 6]
```diff
@@ -0,0 +1,671 @@
...
+
+  //-------------- View surface -------------------------------------------
+  QApplication application(argc,argv);
+  Point4 low4 = K4.lowerBound();
+  Point4 up4  = K4.upperBound();
+  KSpace3 K3;
+  K3.init( Point3( low4[ 0 ], low4[ 1 ], low4[ 2 ] ),
+           Point3( up4 [ 0 ], up4 [ 1 ], up4 [ 2 ] ), true );
+  Viewer3D<Space3,KSpace3> viewer( K3 );
+  viewer.setWindowTitle("simple Volume Viewer");
```
Comment (kerautret) [inline visualisation/3dImplicitSurfaceExtractorBy4DExtension.cpp]: viewer.setWindowTitle("Implicit Surface Extractor By 4D Extension"); ?

File: visualisation/3dImplicitSurfaceExtractorByThickening.cpp [hunk 7]
```diff
@@ -0,0 +1,495 @@
...
+          mesh.addTriangularFace( face_idx[ 1 ], face_idx[ 3 ], face_idx[ 2 ], color );
+        }
+      //mesh.addQuadFace( face_idx[ 0 ], face_idx[ 1 ], face_idx[ 3 ], face_idx[ 2 ], color );
+     }
+  trace.endBlock();
+
+  //-------------- View surface -------------------------------------------
+  QApplication application(argc,argv);
+  Viewer3D<Space3,KSpace3> viewer( K3 );
+  viewer.setWindowTitle("simple Volume Viewer");
```
Comment (kerautret) [inline visualisation/3dImplicitSurfaceExtractorByThickening.cpp]: idem:
 viewer.setWindowTitle("3d Implicit Surface Extractor By Thickening "); ?

File: visualisation/3dImplicitSurfaceExtractorByThickening.cpp [hunk 8]
```diff
@@ -0,0 +1,495 @@
...
+
+/**
+ * @file cubical-complex-collapse.cpp
+ * @ingroup Examples
+ * @author Jacques-Olivier Lachaud (\c jacques-olivier.lachaud@univ-savoie.fr )
+ * Laboratory of Mathematics (CNRS, UMR 5127), University of Savoie, France
+ *
+ * @date 2015/08/28
+ *
+ * An example file named cubical-complex-collapse.cpp.
```
Comment (kerautret) [inline visualisation/3dImplicitSurfaceExtractorByThickening.cpp]: idem

**Comment by kerautret** [comment]
Nice new tools, just minor remarks.

**Comment by kerautret** [comment]
thanks and if you can add a line in the changelog ;)
I will add screen shot on the main readme.md

**Comment by kerautret** [comment]
@JacquesOlivierLachaud you have time to edit for 9.1.0 or later ?

**Comment by JacquesOlivierLachaud** [comment]
I try to do it today.

File: visualisation/3dImplicitSurfaceExtractorBy4DExtension.cpp (hunk 3 above)
Comment (JacquesOlivierLachaud) [inline visualisation/3dImplicitSurfaceExtractorBy4DExtension.cpp]: I do not see the DEC class here ?? derivative is a function in Polynomial.h

File: visualisation/3dImplicitSurfaceExtractorBy4DExtension.cpp (hunk 4 above)
Comment (JacquesOlivierLachaud) [inline visualisation/3dImplicitSurfaceExtractorBy4DExtension.cpp]: Here the problem is that I use both 3D and 4D objects, hence all these typedefs, which is quite heavy I agree.

File: visualisation/3dImplicitSurfaceExtractorBy4DExtension.cpp (hunk 2 above)
Comment (JacquesOlivierLachaud) [inline visualisation/3dImplicitSurfaceExtractorBy4DExtension.cpp]: modified.

File: visualisation/3dImplicitSurfaceExtractorBy4DExtension.cpp (hunk 1 above)
Comment (JacquesOlivierLachaud) [inline visualisation/3dImplicitSurfaceExtractorBy4DExtension.cpp]: Sure.

File: visualisation/3dImplicitSurfaceExtractorBy4DExtension.cpp (hunk 6 above)
Comment (JacquesOlivierLachaud) [inline visualisation/3dImplicitSurfaceExtractorBy4DExtension.cpp]: done.

File: visualisation/3dImplicitSurfaceExtractorByThickening.cpp (hunk 7 above)
Comment (JacquesOlivierLachaud) [inline visualisation/3dImplicitSurfaceExtractorByThickening.cpp]: done.

File: visualisation/3dImplicitSurfaceExtractorByThickening.cpp (hunk 8 above)
Comment (JacquesOlivierLachaud) [inline visualisation/3dImplicitSurfaceExtractorByThickening.cpp]: done.

**Comment by JacquesOlivierLachaud** [comment]
AppVeyor seems to fail because of dynamic boost_programoptions. I had removed in CMakeLists the USE_STATIC_BOOST because it fails on my mac (os el capitan 10.11.2, boost-1.59 installed with macports). Indeed, I do not know why, but macports does not install boost program options static lib (although I have not chosen the variant +no_static). Weird !
I try to put back BOOST_USE_STATIC_LIBS and see what happens.

File: visualisation/3dImplicitSurfaceExtractorBy4DExtension.cpp (hunk 3 above)
Comment (kerautret) [inline visualisation/3dImplicitSurfaceExtractorBy4DExtension.cpp]: ok I see, Xcode show me the ref on dec but was wrong

**Comment by kerautret** [comment]
perfect @JacquesOlivierLachaud, many thanks , I merge

**Pull request merged**